File: 백준/Gold/11689. GCD（n， k） ＝ 1/GCD（n， k） ＝ 1.java
import java.io.*;
import java.util.*;

class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        long n = Long.parseLong(br.readLine());
        br.close();

        long res = n < 3 ? 1 : isPrime(n) ? n - 1 : eulerPhi(n);

        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        bw.write(Long.toString(res));
        bw.close();
    }

    private static boolean isPrime(long n) {
        for (int k = 2; k <= Math.sqrt(n); k++) {
            if (n % k == 0) return false;
        }
        return true;
    }

    private static long eulerPhi(long n) {
        long res = n;
        List<Long> primes = new ArrayList<>();

        for (long k = 2; ; k++) {
            if (n % k == 0 && isPrime(k)) {
                primes.add(k);

                while (n % k == 0) n /= k;

                if (isPrime(n)) break;
            }
        }
        if (n > 1) primes.add(n);

        for (long p : primes) {
            res /= p;
            res *= p - 1;
        }
        return res;
    }
}

File: 백준/Gold/11689. GCD（n， k） ＝ 1/GCD（n， k） ＝ 1.kt
import kotlin.math.sqrt

fun main() = with(System.`in`.bufferedReader()) {
    val n = readLine().toLong()
    val epf = EulerProductFormula()

    val res = epf.query(n)

    with(System.out.bufferedWriter()) {
        append("$res").flush()
        close()
    }
}

class EulerProductFormula {
    fun query(n: Long): Long {
        if (n < 3) return 1

        if (isPrime(n)) return n - 1

        val primes = arrayListOf<Long>()

        for (k in 2..sqrt(n.toDouble()).toLong()) {
            if (n % k > 0) continue

            if (isRelative(k, n / k)) return query(k) * query(n / k)

            if (isPrime(k)) primes.add(k)
        }

        var ret = n

        for (k in primes) {
            ret *= (k - 1)
            ret /= k
        }
        return ret
    }

    fun isRelative(a: Long, b: Long): Boolean {
        for (k in 2..a) {
            if (a % k == 0L && b % k == 0L) return false
        }
        return true
    }

    fun isPrime(n: Long): Boolean {
        for (k in 2..sqrt(n.toDouble()).toInt()) {
            if (n % k == 0L) return false
        }
        return true
    }
}